spec.bs

```diff
@@ -6,57 +6,940 @@ Status: CG-DRAFT
+
+  An [=shared storage database/entry=]'s <dfn for=entry>key</dfn> is a [=string=].
+
+  [=User agents=] may specify the <dfn for=key>maximum length</dfn> of a [=entry/key=].
+
+  Since [=entry/keys=] are used to organize and efficiently retrieve [=shared storage database/entry|entries=], [=entry/keys=] must appear at most once in any given [=shared storage database=].
+
+  An [=shared storage database/entry=]'s <dfn for=entry>value struct</dfn> is a [=struct=] composed of [=string=] <dfn for="value struct">value</dfn> and {{DOMHighResTimeStamp}} <dfn for="value struct">last updated</dfn> (from the [=Unix Epoch=]).
+
+  Note: Some Shared Storage implementations may use SQLite implementations that store timestamps as 64-bit unsigned integers rather than doubles, thereby making it necessary to use legacy {{EpochTimeStamp}}s instead of {{DOMHighResTimeStamp}}s.
+
+  <span class=todo>In Chrome, for the reason given in the note above, we use {{EpochTimeStamp}}s, which are discouraged. How can we reconcile this?</span>
+
+  [=User agents=] may specify the <dfn for=value>maximum length</dfn> of a [=value struct/value=].
+
+  [=User agents=] may specify a <dfn>default entry lifetime</dfn>, the default [=duration=] between when an [=shared storage database/entry=] is [=shared storage database/store an entry in the database|stored=] and when it expires.
+
+  ## The [=Shared Storage Database|Database=] Algorithms ## {#database-algorithms}
+
+  <div algorithm>
+    To <dfn for="shared storage database">store an entry in the database</dfn>, given a [=/browsing context=] |context|, a [=entry/key=] |key|, and a [=value struct/value=] |value|, run the following steps:
+
+    1. If |key| is not a [=string=] or |key|'s [=string/length=] exceeds the [=key/maximum length=], return false.
+    1. If |value| is not a [=string=] or |value|'s [=string/length=] exceeds the [=value/maximum length=], return false.
+    1. Let |environment| be |context|'s [=active window=]'s [=relevant settings object=].
+    1. [=Enqueue the following steps=] on |context|'s associated [=shared storage database/shared storage database queue=]:
+        1. Let |databaseMap| be the result of running [=obtain a shared storage bottle map=] for |environment|.
+        1. If |databaseMap| is failure, then return null.
+        1. Let |valueStruct| be a new [=entry/value struct=].
+        1. Set |valueStruct|'s [=value struct/value=] to |value|.
+        1. Let |unsafeMoment| be the [=wall clock=]'s [=wall clock/unsafe current time=].
+        1. Let |moment| be the result of running [=coarsen time=] algorithm given |unsafeMoment| and [=wall clock=] as input.
+        1. Let |fromUnixEpochTime| be the [=duration from=] the [=Unix epoch=] to |moment|.
+        1. Set |valueStruct|'s [=value struct/last updated=] to |fromUnixEpochTime|.
+        1. [=map/Set=] |databaseMap|[|key|] to |valueStruct|.
+        1. Return true.
```

Comment:
Unfortunately you cannot return a value like this from an enqueued set of steps.  The algorithm synchronously returns undefined before these steps are executed.

There are a couple ways to solve this:

1.  You can make these internal algorithms return promises.  If the promises are always directly returned to javascript through the API layer, then that may be fine.  If you want to use these values inside other algorithms, though, it might get a bit ugly.
2. You can make these internal algorithms synchronous and require that they are only called via the database queue.  The outer API method algorithms then would be responsible for enqueuing the steps that call the internal algorithm.  If you do this you will need to pass in any information needed from the current context; e.g. the environment, etc.

Reply:
Made synchronous, eliminated double checking the inputs